dcrtimed/backend/filesystem/dump.go

```diff
@@ -370,3 +376,479 @@ func (fs *FileSystem) Restore(f *os.File, verbose bool, location string) error {
+				"in empties: %v", ts)
+		}
+		empties[ts] = struct{}{}
+		fmt.Printf("   *** ERROR empty timestamp: %v (%v)\n",
+			ts2dirname(ts), ts)
+		return nil
+	}
+
+	if flushRecord != nil {
+		if options.Verbose {
+			fmt.Printf("   Merkle root    : %x\n",
+				flushRecord.Root)
+			fmt.Printf("   Tx             : %v\n", flushRecord.Tx)
+			fmt.Printf("   Chain timestamp: %v\n",
+				flushRecord.ChainTimestamp)
+			fmt.Printf("   Flush timestamp: %v\n",
+				flushRecord.FlushTimestamp)
+		}
+
+		// 3. Make sure all digests exist in flush record
+		if len(digests) != len(flushRecord.Hashes) {
+			// XXX Dump hashes
+			return fmt.Errorf("   *** ERROR mismatched len: %v %v",
+				len(digests), len(flushRecord.Hashes))
+		}
+		for _, v := range flushRecord.Hashes {
+			hash := hex.EncodeToString(v[:])
+			if options.PrintHashes {
+				fmt.Printf("     Hashes       : %v\n", hash)
+			}
+
+			if _, ok := digests[hash]; !ok {
+				return fmt.Errorf("   *** ERROR unknown hash: %v",
+					hash)
+			}
+		}
+
+		// 3.1 Recreate merkle and verify it
+		root := merkle.Root(flushRecord.Hashes)
+		if !bytes.Equal(root[:], flushRecord.Root[:]) {
+			return fmt.Errorf("   *** ERROR mismatched merkle "+
+				"root: %x %x", *root, flushRecord.Root)
+		}
+
+		// 3.2 verify that all digests exist in global db and verify
+		// timestamp points to the correct container.
+		for _, v := range flushRecord.Hashes {
+			gdbts, err := fs.db.Get(v[:], nil)
+			if err != nil {
+				return fmt.Errorf("   *** ERROR not found in "+
+					"db: %x", v)
+			}
+			dbts := int64(binary.LittleEndian.Uint64(gdbts))
+			if dbts != ts {
+				return fmt.Errorf("   *** ERROR timestamp "+
+					"mismatch: %v %v", dbts, ts)
+			}
+		}
+
+		// 3.3 Verify merkle root in tx
+		u := "https://alpha.dcrdata.org/api/tx/" +
+			flushRecord.Tx.String() + "/out"
+		r, err := http.Get(u)
+		if err != nil {
+			return fmt.Errorf("   *** ERROR invalid dcrdata "+
+				"answer: %v", err)
+		}
+		var txOuts []types.TxOut
+		d := json.NewDecoder(r.Body)
+		if err := d.Decode(&txOuts); err != nil {
+			return err
+		}
+
+		var done bool
+		for _, v := range txOuts {
+			if !types.IsNullDataScript(v.ScriptPubKeyDecoded.Type) {
+				continue
+			}
+			script, err := hex.DecodeString(v.ScriptPubKeyDecoded.Hex)
+			if err != nil {
+				// XXX bad error, fix
+				return fmt.Errorf("   *** ERROR invalid "+
+					"dcrdata script: %v", err)
+			}
+			data, err := txscript.PushedData(script)
+			if err != nil {
+				// XXX bad error, fix
+				return fmt.Errorf("   *** ERROR invalid "+
+					"script: %v", err)
+			}
+			if !bytes.Equal(data[0], flushRecord.Root[:]) {
+				continue
+			}
+
+			// Everything is cool so mark it and break out
+			done = true
+			break
+		}
+		if !done {
+			return fmt.Errorf("   *** ERROR merkle root not "+
+				"found: tx %v merkle %x", flushRecord.Tx,
+				flushRecord.Root)
+		}
+
+		// We are done
+		return nil
+	}
+
+	// 3. Make sure timestamps do not exist in the global database
```

Comment:
Isn't this `2.` according to the function declaration comment, not `3.`?

Reply:
I flipped the code around and that got massacred. Fixed.

Reply:
Resolved.